In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# creating a path to load data fames to python 
path = r"C:\Users\pshan\Downloads\InstaCart Basket Analysis"

In [3]:
# loading orders_products_combined dataset into jupytor
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [4]:
# slicing a part of data frame 
df = df_ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 18)

In [6]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days       Average orders   
1  Regularly busy      Least busy days       Average orders   
2  Regularly busy      Least busy days           Most orders  
3      Least busy      Least busy days       Average orders   
4      Least busy      Least busy days           Most orders

In [7]:
# grouping by on product_name
df.groupby('product_name')

In [8]:
# getting the mean order_numbers grouped by derpartment_id
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# getting the mean, min and max order_numbers grouped by derpartment_id
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [11]:
# creating a new column max_order to store the maximum number of order a user has placed
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
df_ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
..       ...      ...           ...                 ...                ...   
95   3226575      360             1                   5                 12   
96   1469869      377             3                   5                 17   
97   1927023      387             2                   4                 10   
98    858092      420             4                   1                 19   
99    119314      420             5                   3                 19   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
..                     ...         ...                ...        ...   
95                     NaN         196                  1          0   
96                     3.0         196                  9          0   
97                    22.0         196                  3          0   
98                    30.0         196                  2          0   
99                    16.0         196                  2          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
..          ...       ...            ...     ...    ...                ...   
95         Soda        77              7     9.0   both  Mid-range product   
96         Soda        77              7     9.0   both  Mid-range product   
97         Soda        77              7     9.0   both  Mid-range product   
98         Soda        77              7     9.0   both  Mid-range product   
99         Soda        77              7     9.0   both  Mid-range product   

       busiest_day         busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy days       Average orders          10  
1   Regularly busy      Least busy days       Average orders          10  
2   Regularly busy      Least busy days           Most orders         10  
3       Least busy      Least busy days       Average orders          10  
4       Least busy      Least busy days           Most orders         10  
..             ...                  ...                   ...        ...  
95  Regularly busy  Regularly busy days           Most orders          3  
96  Regularly busy  Regularly busy days       Average orders           3  
97      Least busy      Least busy days           Most orders          8  
98  Regularly busy         Busiest days       Average orders          22  
99  Regularly busy      Least busy days       Average orders          22  

[100 rows x 19 columns]

In [13]:
#  To display max rows
pd.options.display.max_rows = None

In [14]:
df_ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

### to set the flag in loyaity flag column

In [15]:
# if the max order by customer is more then 40 the customer will be flagged as 'loyal customer' 
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
# if the max order by customer is more then 10 and upto 40 the customer will be flagged as 'Regular customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
# if the max order by customer is less then 10 the customer will be flagged as 'New customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# getting the stats on this new column
df_ords_prods_merge['loyalty_flag'].value_counts( dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [19]:
# getting only required column to verify the results
df_ords_prods_merge[['user_id', 'loyalty_flag', 'max_order','order_number']].head(60)

user_id      loyalty_flag  max_order  order_number
0         1      New customer         10             1
1         1      New customer         10             2
2         1      New customer         10             3
3         1      New customer         10             4
4         1      New customer         10             5
5         1      New customer         10             6
6         1      New customer         10             7
7         1      New customer         10             8
8         1      New customer         10             9
9         1      New customer         10            10
10       15  Regular customer         22            15
11       15  Regular customer         22            17
12       15  Regular customer         22            18
13       15  Regular customer         22            21
14       15  Regular customer         22            22
15       19      New customer          9             2
16       19      New customer          9             5
17       19      New customer          9             7
18       21  Regular customer         33            10
19       31  Regular customer         20            10
20       31  Regular customer         20            17
21       43  Regular customer         11             6
22       43  Regular customer         11             9
23       52  Regular customer         27             1
24       52  Regular customer         27             2
25       52  Regular customer         27             3
26       52  Regular customer         27             5
27       52  Regular customer         27             7
28       52  Regular customer         27             9
29       52  Regular customer         27            10
30       52  Regular customer         27            11
31       52  Regular customer         27            12
32       52  Regular customer         27            15
33       52  Regular customer         27            17
34       52  Regular customer         27            18
35       52  Regular customer         27            19
36       52  Regular customer         27            21
37       67  Regular customer         24             1
38       67  Regular customer         24             2
39       67  Regular customer         24             3
40       67  Regular customer         24             4
41       67  Regular customer         24             5
42       67  Regular customer         24             6
43       67  Regular customer         24             8
44       67  Regular customer         24             9
45       67  Regular customer         24            10
46       67  Regular customer         24            11
47       67  Regular customer         24            12
48       67  Regular customer         24            13
49       67  Regular customer         24            14
50       67  Regular customer         24            16
51       67  Regular customer         24            18
52       67  Regular customer         24            19
53       67  Regular customer         24            21
54       67  Regular customer         24            22
55       67  Regular customer         24            23
56       81      New customer          7             3
57       81      New customer          7             5
58       82  Regular customer         19             1
59       82  Regular customer         19             3

In [20]:
# getting the mean order_numbers grouped by derpartment_id for the entire dataset
df_ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

In [21]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [22]:
df['department_id'].value_counts( dropna = False)

department_id
4     611084
16    169624
20    108866
7      44710
19     38095
13     22656
14      3751
17      1214
Name: count, dtype: int64

In [23]:
df_ords_prods_merge['department_id'].value_counts( dropna = False)

department_id
4     9479291
16    5398747
19    2887550
7     2688123
1     2234743
13    1875369
3     1172428
15    1068058
20    1051249
9      866627
17     738666
12     708927
14     703033
11     447572
18     423802
6      269253
5      153696
8       97716
21      69145
2       36291
10      34573
Name: count, dtype: int64

#### the subset of the entire data set df only included the first 1000000 rows and thus did not had every single department id, moreover the included department ids has only got a subset of data thats why the mean in df is more than the mean in entire dataframe, only the mean of department_id of 17 in df is lesser than the mean in entire data frame, which makes complete sense as in df there were only 1214 rows associated with department_id 17 whereas in actual the row count was 738666

### basic statistics of the product prices for each loyalty category

In [24]:
df_ords_prods_merge['loyalty_flag'].value_counts( dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [26]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773487  1.0  25.0
New customer      7.801077  1.0  25.0
Regular customer  7.798145  1.0  25.0

####  the mean prices of loyal customer are the least as compared to the others, the mean prices of orders by new customer are the highest

In [27]:
# exporting the df
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_01.pkl'))